pkg/urbit/jets/e/en_json.c

```diff
@@ -0,0 +1,438 @@
+  const c3_y _JSON_NEWLINE[] = "\\n";
+  const c3_y _JSON_DOQ[] = "\\\"";
+  const c3_y _JSON_BAS[] = "\\\\";
+  const c3_y *_JSON_UNICODES[] = {
+    "\\u0000",  // 0
+    "\\u0001",  // 1
+    "\\u0002",  // 2
+    "\\u0003",  // 3
+    "\\u0004",  // 4
+    "\\u0005",  // 5
+    "\\u0006",  // 6
+    "\\u0007",  // 7
+    "\\u0008",  // 8
+    "\\u0009",  // 9
+    "\\u000a",  // 10
+    "\\u000b",  // 11
+    "\\u000c",  // 12
+    "\\u000d",  // 13
+    "\\u000e",  // 14
+    "\\u000f",  // 15
+    "\\u0010",  // 16
+    "\\u0011",  // 17
+    "\\u0012",  // 18
+    "\\u0013",  // 19
+    "\\u0014",  // 20
+    "\\u0015",  // 21
+    "\\u0016",  // 22
+    "\\u0017",  // 23
+    "\\u0018",  // 24
+    "\\u0019",  // 25
+    "\\u001a",  // 26
+    "\\u001b",  // 27
+    "\\u001c",  // 28
+    "\\u001d",  // 29
+    "\\u001e",  // 30
+    "\\u001f",  // 31
+  };
+  const c3_y _JSON_DEL[] = "\\u007f";
+
+/* forward declarations
+*/
+
+  json_buffer*
+  _serialize(json_buffer*, u3_noun);
+
+/* core jet logic
+*/
+
+  json_buffer*
+  _init_buffer(json_buffer *buf_u)
+  {
+    buf_u->buf_y = (c3_y*)u3a_malloc(SERIALIZE_BUFFER_INIT_SIZE);
+    buf_u->len_w = 0;
+    buf_u->siz_w = SERIALIZE_BUFFER_INIT_SIZE;
+  }
+
+  void
+  _close_buffer(json_buffer *buf_u)
+  {
+    u3a_free(buf_u->buf_y);
+  }
+
+  json_buffer*
+  _encode_fail(json_buffer *buf_u)
+  {
+    _close_buffer(buf_u);
+    return NULL;
+  }
+
+  json_buffer*
+  _resize_buffer(json_buffer *buf_u)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+
+    buf_u->siz_w += SERIALIZE_BUFFER_INCR_SIZE;
+    buf_u->buf_y = (c3_y*)u3a_realloc(buf_u->buf_y, buf_u->siz_w);
+
+    return buf_u;
+  }
+
+  json_buffer*
+  _append_char(json_buffer *buf_u, c3_y c_y)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+
+    if ((buf_u->len_w + 1) >= buf_u->siz_w) {
+      buf_u = _resize_buffer(buf_u);
+    }
+
+    buf_u->buf_y[(buf_u->len_w)++] = c_y;
+    return buf_u;
+  }
+
+  json_buffer*
+  _append_text(json_buffer *buf_u, const c3_y *buf_y, c3_w len_w)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+    
+    while ((buf_u->len_w + len_w) >= buf_u->siz_w) {
+      buf_u = _resize_buffer(buf_u);
+    }
+
+    memcpy(&(buf_u->buf_y[buf_u->len_w]), buf_y, len_w);
+    buf_u->len_w += len_w;
+    return buf_u;
+  }
+
+  json_buffer*
+  _serialize_null(json_buffer *buf_u)
+  {
+    _append_text(buf_u, _JSON_NULL, 4);
+  }
+
+  json_buffer*
+  _serialize_true(json_buffer *buf_u)
+  {
+    _append_text(buf_u, _JSON_TRUE, 4);
+  }
+
+  json_buffer*
+  _serialize_false(json_buffer *buf_u)
+  {
+    _append_text(buf_u, _JSON_FALSE, 5);
+  }
+
+  json_buffer*
+  _serialize_loobean(json_buffer *buf_u, u3_noun a)
+  {
+    if ( buf_u == NULL ) {
+      return NULL;
+    }
+
+    if ( a == c3y ) {
+      return _serialize_true(buf_u);
+    }
+    else if ( a == c3n ) {
+      return _serialize_false(buf_u);
+    }
+    else {
+      return _encode_fail(buf_u);
+    }
```

Comment:
This should be a switch on `a`, with a `u3m_bail(c3__exit)` in the default case. (This is the standard idiom for consuming loobeans in jets.)

Reply:
Done